synapse/cores/common.py

```diff
@@ -2858,6 +2890,139 @@ def snapTufosByDark(self, name, valu=None, mintime=None, maxtime=None, limit=Non
+        return None
+
+    def _hasBlobValu(self, key):  # pragma: no cover
+        self.log(logging.ERROR, mesg='Core does not implement _hasBlobValue', name='_hasBlobValue')
+        return None
+
+    def _delBlobValu(self, key):  # pragma: no cover
+        self.log(logging.ERROR, mesg='Core does not implement _delBlobValu', name='_delBlobValu')
+        return None
+
+    def _getBlobKeys(self):
+        self.log(logging.ERROR, mesg='Core does not implement _getBlobKeys', name='_getBlobKeys')
+        return None
+
+    def _getCoreType(self):  # pragma: no cover
+        raise s_common.NoSuchImpl(name='getCoreType', mesg='Core does not implement getCoreType')
+
+    def getCoreType(self):
+        return self._getCoreType()
+
+    # TODO: Wrap this in a userauth layer
+    def getBlobValu(self, key, default=None):
+        '''
+        Get a value from the blob key/value (KV) store.
+
+        This resides below the tufo storage layer and is Cortex implementation
+        dependent. In purely memory backed cortexes, this KV store may not be
+        persistent, even if the tufo-layer is persistent, through something
+        such as the savefile mechanism.
+
+        Notes:
+            Data which is retrieved from the KV store is msgpacked, so caveats
+            with that apply.
+
+        Args:
+            key (str): Value to retrieve
+            default: Value returned if the key is not present in the blob store.
+
+        Returns:
+            The value from the KV store or the default valu (None).
+
+        '''
+        buf = self._getBlobValu(key)
+        if buf is None:
+            self.log(logging.WARNING, mesg='Requested key not present in blob store, returning default', name=key)
+            return default
+        return s_common.msgunpack(buf)
+
+    def getBlobKeys(self):
+        '''
+        Get a list of keys in the blob key/value store.
+
+        Returns:
+            list: List of keys in the store.
+        '''
+        return self._getBlobKeys()
+
+    # TODO: Wrap this in a userauth layer
+    def setBlobValu(self, key, valu):
+        '''
+        Set a value from the blob key/value (KV) store.
+
+        This resides below the tufo storage layer and is Cortex implementation
+        dependent. In purely memory backed cortexes, this KV store may not be
+        persistent, even if the tufo-layer is persistent, through something
+        such as the savefile mechanism.
+
+        Notes:
+            Data which is stored in the KV store is msgpacked, so caveats with
+            that apply.
+
+        Args:
+            key (str): Name of the value to store.
+            valu: Value to store in the KV store.
+
+        Returns:
+            The input value, unchanged.
+        '''
+        buf = s_common.msgenpack(valu)
+        self.savebus.fire('syn:core:blob:set', key=key, valu=buf)
+        self._setBlobValu(key, buf)
```

Comment:
we should call _setBlobValu before fire() so consumers of the event can assume that the storage layer has already done it's thing.